salt/_states/metalk8s_kubernetes.py

```diff
@@ -0,0 +1,163 @@
+    return __virtualname__
+
+
+def object_absent(name, manifest=None, **kwargs):
+    """Ensure that the object is absent.
+
+    Arguments:
+        name (str): Path to a manifest yaml file or just a name
+        manifest (dict): Manifest content
+    """
+    ret = {'name': name, 'changes': {}, 'result': True, 'comment': ''}
+
+    # Only pass `name` if we have no manifest
+    name_arg = None if manifest else name
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name=name_arg, manifest=manifest,
+        **kwargs)
+
+    if obj is None:
+        ret['comment'] = 'The object does not exist'
+        return ret
+
+    if __opts__['test']:
+        ret['result'] = None
+        ret['comment'] = 'The object is going to be deleted'
+        return ret
+
+    result = __salt__['metalk8s_kubernetes.delete_object'](
+        name=name_arg, manifest=manifest,
+        **kwargs)
+
+    if result is not None:
+        ret['changes'] = {'old': 'present', 'new': 'absent'}
+        ret['comment'] = 'The object was deleted'
+    else:
+        # This happens if the DELETE call fails with a 404 status
+        ret['comment'] = 'The object does not exist'
+
+    return ret
+
+
+def object_present(name, manifest=None, **kwargs):
+    """Ensure that the object is present.
+
+    Arguments:
+        name (str): Path to a manifest yaml file
+                    or just a name if manifest provided
+        manifest (dict): Manifest content
+    """
+    ret = {'name': name, 'changes': {}, 'result': True, 'comment': ''}
+
+    # Only pass `name` if we have no manifest
+    name_arg = None if manifest else name
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name=name_arg, manifest=manifest,
+        **kwargs
+    )
+    if __opts__['test']:
+        ret['result'] = None
+        ret['comment'] = 'The object is going to be {}'.format(
+            'created' if obj is None else 'replaced'
+        )
+        return ret
+
+    if obj is None:
+        __salt__['metalk8s_kubernetes.create_object'](
+            name=name_arg, manifest=manifest,
+            **kwargs
+        )
+        ret['changes'] = {'old': 'absent', 'new': 'present'}
+        ret['comment'] = 'The object was created'
+
+        return ret
+
+    # TODO: Attempt to handle idempotency as much as possible here, we don't
+    #       want to always replace if nothing changed. Currently though, we
+    #       don't know how to achieve this, since some fields may be set by
+    #       api-server or by the user without us being able to distinguish
+    #       them.
+    new = __salt__['metalk8s_kubernetes.replace_object'](
+        name=name_arg, manifest=manifest,
+        **kwargs
+    )
+    diff = __utils__['dictdiffer.recursive_diff'](obj, new)
+    ret['changes'] = diff.diffs
+    ret['comment'] = 'The object was replaced'
+
+    return ret
+
+
+def object_updated(name, manifest=None, **kwargs):
```

Comment:
Yes I know btw you only need to have a part of the manifest so in manifest: `metadata.name` `kind` and `apiVersion` to find the object and then only the field you want to patch. 
But indeed I want to add something like for `delete` to be able to use only `name` `kind` `apiVersion` and `patch` directly in the state.

I will do it

Reply:
> you only need to have a part of the manifest so in manifest: `metadata.name` `kind` and `apiVersion` to find the object and then only the field you want to patch

That's not super obvious or intuitive (not saying it's your fault or anything, we do the same for `present` and `absent`).

What we could do, for those states, is have something easily written by humans, and let the renderer adapt manifests to those signatures, instead of doing the opposite.
Here's what I imagine:

```yaml
Ensure some object is present with its manifest:
  metalk8s_kubernetes.object_present:
    name: some-object
    manifest:
      apiVersion: v1
      kind: Pod
      metadata: [...]
      spec: [...]

Ensure some object is present with a source file:
  metalk8s_kubernetes.object_present:
    name: some-object
    source: /path/to/manifest.yaml
    template: jinja
    context:
      image: {{ build_image('my-image') }}

Ensure some object is absent:
  metalk8s_kubernetes.object_absent:
    name: some-object
    namespace: some-namespace
    apiVersion: v1
    kind: Pod

Have some object updated with strategic merge:
  metalk8s_kubernetes.object_updated:
    name: some-object
    namespace: some-namespace
    apiVersion: v1
    kind: Pod
    patch:
      metadata:
        annotations:
          metalk8s.scality.com/my-annotation: my-value

Have some object updated with JSON patch:
  metalk8s_kubernetes.object_updated:
    name: some-object
    namespace: some-namespace
    apiVersion: v1
    kind: Pod
    patch:
      - op: add
        path: /spec/containers
        value:
          name: my-sidecar
          [...]
```

Note that the client chooses between `Content-Type: application/json-patch+json` and `Content-Type: application/strategic-merge-patch+json` based on the type of `patch` here: if it's a `list`, then it will use JSON Patch, otherwise it chooses [Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md), which is a Merge Patch but much smarter based on the source Go structs and OpenAPI annotations there. See the link for more info on what can be done to override these annotations using `$patch` directives inside the `patch`.